### PR TITLE
EC-2026: Verify SBOM attestations with signing identities

### PR DESCRIPTION
**Pull request by joejstuart**
## Summary

- verify attached CycloneDX and SPDX SBOM attestations with signing_identities.sbom
- fail closed when the SBOM signing identity is absent or invalid
- update unit and acceptance coverage for the verified-attestation boundary

## Testing

- make ci

Jira: https://redhat.atlassian.net/browse/EC-2026

### CODE REVIEW

File: policy/lib/sbom/sbom.rego [hunk 1]
```diff
@@ -514,21 +527,31 @@ rule_data_errors contains error if {
...
 
+_sbom_signing_identity_configured if {
+	is_object(_sbom_signing_identity)
+	count(_sbom_signing_identity) > 0
+}
+
 # Collect SBOM signature verification errors for observability. Fail-closed
 # exclusion happens in the verified_* wrappers; this surfaces WHY an SBOM
 # was excluded, using the failures API to avoid duplicate builtin calls.
 signature_verification_errors contains error if {
```
[the remark this answers is not in the document]
Comment (joejstuart) [reply policy/lib/sbom/sbom.rego:574]: No change planned. When no SBOM identity is configured, signature verification is intentionally not attempted, so there is no verification failure to report. Invalid configured identities are surfaced through optional_identity_rule_data_errors, while absence remains valid for PipelineRun-derived SBOMs.

File: acceptance/policy/sbom_proxy/lib/sbom_acceptance_shim.rego [hunk 2]
```diff
@@ -0,0 +1,11 @@
+package lib.sbom
```
[the remark this answers is not in the document]
Comment (joejstuart) [reply acceptance/policy/sbom_proxy/lib/sbom_acceptance_shim.rego:1]: Fixed in f2b08647 by moving the shim to acceptance/policy/sbom_proxy and loading only that scoped path from the proxy scenarios. Runtime verification remains covered by focused unit tests; the proxy acceptance fixtures intentionally test downstream behavior using captured input without image references.

File: policy/lib/sbom/sbom.rego [hunk 3]
```diff
@@ -514,21 +527,31 @@ rule_data_errors contains error if {
 _sbom_identity_name := "sbom"
 
 # _sbom_signing_identity is the named entry in the signing_identities rule
-# data used to verify SBOMs discovered via OCI referrers and image-tag refs.
-# It is undefined (fail-closed, silent) when no such identity is configured.
+# data used to verify SBOM attestations, OCI referrers, and image-tag refs.
+# Without a configured identity it defaults to an empty object; the configured
+# guard prevents verification and no SBOMs from those sources are accepted.
+default _sbom_signing_identity := {}
```
[the remark this answers is not in the document]
Comment (joejstuart) [reply policy/lib/sbom/sbom.rego:554]: No change planned. A non-object identity fails the configured guard and therefore fails closed, while optional_identity_rule_data_errors reports the invalid shape to consumers. This is the intended separation between gating and configuration diagnostics.

File: policy/lib/sbom/sbom.rego [hunk 4]
```diff
@@ -48,12 +51,20 @@ _cyclonedx_sboms_from_image contains sbom if {
...
 	statement := att.statement
 	statement.predicateType == "https://spdx.dev/Document"
 }
 
+# Attached SBOM attestations are verified against the named "sbom" signing
+# identity.
+_verified_sbom_attestations contains attestation if {
+	_sbom_signing_identity_configured
+	verification := ec.sigstore.verify_attestation(input.image.ref, _sbom_signing_identity)
+	some attestation in verification.attestations
```
[the remark this answers is not in the document]
Comment (joejstuart) [reply policy/lib/sbom/sbom.rego:80]: Fixed in f2b08647. The rule now requires object.get(verification, "success", false) == true before consuming attestations. A regression test covers success:false with a non-empty attestations array.

File: policy/lib/sbom/sbom.rego [hunk 5]
```diff
@@ -48,12 +64,21 @@ _cyclonedx_sboms_from_image contains sbom if {
...
+# Verify attached SBOM attestations once, using the named "sbom" signing
+# identity, before either format-specific rule consumes them. With no configured
+# identity this set remains empty, so attached attestations are excluded.
+_verified_sbom_attestations contains attestation if {
+	_sbom_signing_identity_configured
+	verification := ec.sigstore.verify_attestation(input.image.ref, _sbom_signing_identity)
+	some attestation in verification.attestations
+}
+
 _spdx_sboms_from_pipelinerun contains sbom if {
```
[the remark this answers is not in the document]
Comment (joejstuart) [reply policy/lib/sbom/sbom.rego:83]: Documented in f2b08647. This source restriction is intentional for EC-2026: restoring an input-derived clause would reintroduce the unverified trust path the story removes. The design document now explicitly explains the registry requirement and offline/input-only limitation.

File: policy/lib/sbom/sbom.rego [hunk 6]
```diff
@@ -48,12 +51,20 @@ _cyclonedx_sboms_from_image contains sbom if {
...
-	some att in input.attestations
+_spdx_sboms_from_attestations contains statement.predicate if {
+	some att in _verified_sbom_attestations
 	statement := att.statement
 	statement.predicateType == "https://spdx.dev/Document"
 }
 
+# Attached SBOM attestations are verified against the named "sbom" signing
+# identity.
+_verified_sbom_attestations contains attestation if {
```
[the remark this answers is not in the document]
Comment (joejstuart) [reply policy/lib/sbom/sbom.rego:76]: Fixed in f2b08647. The verification success gate is now explicit, and the mixed-result regression test confirms that partial results are rejected.

File: policy/lib/sbom/sbom.rego [hunk 7]
```diff
@@ -27,11 +27,14 @@ spdx_sboms := [s | some s in _spdx_sboms]
...
-_spdx_sboms := (_spdx_sboms_from_input | _spdx_sboms_from_pipelinerun) | _spdx_sboms_from_image
+_cyclonedx_sboms := _cyclonedx_sboms_without_image | _cyclonedx_sboms_from_image
+_cyclonedx_sboms_without_image := _cyclonedx_sboms_from_attestations | _cyclonedx_sboms_from_pipelinerun
 
-_cyclonedx_sboms_from_input contains statement.predicate if {
-	some att in input.attestations
+_spdx_sboms := _spdx_sboms_without_image | _spdx_sboms_from_image
+_spdx_sboms_without_image := _spdx_sboms_from_attestations | _spdx_sboms_from_pipelinerun
+
+_cyclonedx_sboms_from_attestations contains statement.predicate if {
```
[the remark this answers is not in the document]
Comment (joejstuart) [reply policy/lib/sbom/sbom.rego:49]: The trust narrowing is intentional and is the core EC-2026 behavior, so we are not restoring unconditional input.attestations or adding an effective_on policy rule. Commit f2b08647 now documents the migration explicitly in design/sigstore-verification.md and example/data/rule_data.yml. A missing SBOM identity remains valid for PipelineRun-derived SBOM users, so a global missing-identity warning would be misleading.

File: example/data/rule_data.yml [hunk 8]
```diff
@@ -38,9 +38,10 @@ rule_data:
     rh-release:
       public_key: "k8s://openshift-pipelines/release-signing-key"
       ignore_rekor: true
-    # Add an "sbom" identity to gate SBOM discovery via OCI referrers and
-    # image-tag refs: SBOMs are only accepted when their signatures verify
-    # against it, and excluded (fail-closed) when it is absent. Left
+    # Add an "sbom" identity to verify attached SBOM attestations and gate
```
[the remark this answers is not in the document]
Comment (joejstuart) [reply example/data/rule_data.yml]: Fixed in f2b08647. The example now states that the SBOM identity is required, explains that input attestations were previously exposed without policy-side verification, and describes the new fail-closed behavior.

File: policy/lib/sbom/sbom_test.rego [hunk 9]
```diff
@@ -811,6 +870,30 @@ test_no_verification_errors_when_no_opts if {
```
[the remark this answers is not in the document]
Comment (joejstuart) [reply policy/lib/sbom/sbom_test.rego:897]: No change planned. The success mocks are named by the format-specific payload they return, while the failure mock is named by outcome. Renaming them would be style-only churn and would not improve test behavior or coverage.

File: policy/lib/sbom/sbom.rego (hunk 6 above)
[the remark this answers is not in the document]
Comment (joejstuart) [reply policy/lib/sbom/sbom.rego:76]: No change planned. The current location keeps the two format-specific attestation consumers together with their shared verifier before the PipelineRun and image-discovery sections. Moving it would be organizational churn without changing behavior or clarity materially.

File: acceptance/policy/sbom_proxy/lib/sbom_acceptance_shim.rego [hunk 10]
```diff
@@ -0,0 +1,11 @@
...
+
+import rego.v1
+
+# The SBOM proxy scenarios use static policy-input samples captured after CLI
+# validation. Those samples do not retain an image reference, so expose their
+# embedded SBOMs to the downstream proxy rules exercised by these scenarios.
+# Runtime signature verification is covered by policy/lib/sbom/sbom_test.rego.
+_verified_sbom_attestations contains attestation if {
```
[the remark this answers is not in the document]
Comment (joejstuart) [reply acceptance/policy/sbom_proxy/lib/sbom_acceptance_shim.rego:10]: Partially addressed in f2b08647 by isolating the shim to the sbom_proxy acceptance path. These fixtures intentionally lack image references and exercise downstream proxy behavior, so they cannot perform live signature verification. The fail-closed boundary is covered by focused unit tests for missing identity, verification failure, and now mixed partial results.

[the remark this answers is not in the document]
Comment (joejstuart) [reply acceptance/policy/sbom_proxy/lib/sbom_acceptance_shim.rego:10]: Addressed in f2b08647 by moving the shim to acceptance/policy/sbom_proxy/lib/sbom_acceptance_shim.rego and loading that scenario-specific path only from sbom_proxy.feature. This keeps it out of the generic acceptance policy path and makes its test-only scope explicit.

File: acceptance/policy/sbom_proxy/lib/sbom_acceptance_shim.rego (hunk 2 above)
[the remark this answers is not in the document]
Comment (joejstuart) [reply acceptance/policy/sbom_proxy/lib/sbom_acceptance_shim.rego:1]: Fixed in f2b08647. The file is now named sbom_acceptance_shim.rego under the scenario-specific acceptance/policy/sbom_proxy path, and its comment explicitly identifies its scope.

File: policy/lib/sbom/sbom.rego (hunk 6 above)
[the remark this answers is not in the document]
Comment (joejstuart) [reply policy/lib/sbom/sbom.rego:76]: No change planned. This is a repository-internal test seam, not a supported downstream API. Stubbing the public SBOM arrays would also bypass the format classification and composition that these consumer tests need to exercise; overriding the private verification boundary keeps those paths under test.

File: policy/lib/sbom/sbom.rego [hunk 11]
```diff
@@ -48,12 +64,21 @@ _cyclonedx_sboms_from_image contains sbom if {
...
 }
 
+# Verify attached SBOM attestations once, using the named "sbom" signing
+# identity, before either format-specific rule consumes them. With no configured
+# identity this set remains empty, so attached attestations are excluded.
+_verified_sbom_attestations contains attestation if {
+	_sbom_signing_identity_configured
+	verification := ec.sigstore.verify_attestation(input.image.ref, _sbom_signing_identity)
+	some attestation in verification.attestations
+}
```
[the remark this answers is not in the document]
Comment (joejstuart) [reply policy/lib/sbom/sbom.rego:81]: Fixed in f2b08647. The policy now requires the overall success flag, and the new regression test proves that a non-empty attestations array is rejected when success is false.

File: design/sigstore-verification.md [hunk 12]
```diff
@@ -71,7 +71,8 @@ each signing identity is self-contained.
 | `lib/intoto/trust.rego` | `object.get(verification, "success", false) == true` | Boolean pass/fail gate |
 | `base_image_registries.rego` | `object.get(info, "success", false) == true` | Boolean pass/fail gate |
 | `lib/oci/oci.rego` | `object.get(info, "errors", [])` per referrer, filtered by `errors == []` | Gate a set of OCI referrers/tag refs to the signature-verified subset (fail-closed) |
-| `lib/sbom/sbom.rego` | `some raw_error in result.errors` over the failures set | Surface why each discovered SBOM was excluded, as warnings |
+| `lib/sbom/sbom.rego` | `some attestation in verification.attestations` | Accept attached SBOM attestations verified with the named `sbom` identity |
```
[the remark this answers is not in the document]
Comment (joejstuart) [reply design/sigstore-verification.md:74]: Fixed in f2b08647. The design document now describes the trust narrowing, fail-closed semantics, registry-verifiability requirement, offline/input-only behavior, and the unaffected PipelineRun path.

File: policy/lib/sbom/sbom.rego [hunk 13]
```diff
@@ -514,21 +527,31 @@ rule_data_errors contains error if {
...
-# data used to verify SBOMs discovered via OCI referrers and image-tag refs.
-# It is undefined (fail-closed, silent) when no such identity is configured.
+# data used to verify SBOM attestations, OCI referrers, and image-tag refs.
+# Without a configured identity it defaults to an empty object; the configured
+# guard prevents verification and no SBOMs from those sources are accepted.
+default _sbom_signing_identity := {}
+
 _sbom_signing_identity := sigstore.named_identity(_sbom_identity_name)
 
+_sbom_signing_identity_configured if {
```
[the remark this answers is not in the document]
Comment (joejstuart) [reply policy/lib/sbom/sbom.rego:558]: No change planned. This predicate answers only whether a non-empty identity object is configured. The shared optional_identity_rule_data_errors helper performs full shape and verification-method validation, and invalid options fail closed in the builtin. Duplicating that validation in the guard would create two sources of truth.

File: policy/lib/sbom/sbom.rego (hunk 3 above)
[the remark this answers is not in the document]
Comment (joejstuart) [reply policy/lib/sbom/sbom.rego:554]: No change planned. The default empty object is intentional so the value remains defined, while _sbom_signing_identity_configured explicitly prevents builtin evaluation and enforces fail-closed behavior. The adjacent comment documents this reason.

File: policy/lib/sbom/sbom.rego [hunk 14]
```diff
@@ -48,12 +64,22 @@ _cyclonedx_sboms_from_image contains sbom if {
...
 
-_spdx_sboms_from_input contains statement.predicate if {
-	some att in input.attestations
+_spdx_sboms_from_attestations contains statement.predicate if {
+	some att in _verified_sbom_attestations
 	statement := att.statement
 	statement.predicateType == "https://spdx.dev/Document"
 }
 
+# Verify attached SBOM attestations once, using the named "sbom" signing
```
[the remark this answers is not in the document]
Comment (joejstuart) [reply policy/lib/sbom/sbom.rego:73]: No change planned. Identity absence is intentionally silent because the SBOM identity is optional for PipelineRun-derived SBOM users. The new observability branch in b4220fc1 is deliberately limited to configured identities whose verification actually fails.

File: acceptance/policy/sbom_proxy/lib/sbom_acceptance_shim.rego [hunk 15]
```diff
@@ -0,0 +1,12 @@
+package lib.sbom
```
[the remark this answers is not in the document]
Comment (joejstuart) [reply acceptance/policy/sbom_proxy/lib/sbom_acceptance_shim.rego:1]: No change planned. The shim is isolated under acceptance/policy/sbom_proxy, loaded only by that feature, explicitly named as an acceptance shim, and acceptance policy is not shipped in production bundles. Sharing package lib.sbom is intentional so these static fixtures can exercise downstream production rules; adding a sentinel would add test configuration without changing the explicit-load trust boundary.

File: policy/lib/sbom/sbom.rego [hunk 16]
```diff
@@ -514,21 +548,31 @@ rule_data_errors contains error if {
...
 
+_sbom_signing_identity_configured if {
+	is_object(_sbom_signing_identity)
+	count(_sbom_signing_identity) > 0
+}
+
 # Collect SBOM signature verification errors for observability. Fail-closed
 # exclusion happens in the verified_* wrappers; this surfaces WHY an SBOM
 # was excluded, using the failures API to avoid duplicate builtin calls.
 signature_verification_errors contains error if {
```
[the remark this answers is not in the document]
Comment (joejstuart) [reply policy/lib/sbom/sbom.rego:574]: Fixed in b4220fc1. Configured attached-attestation verification failures now contribute to signature_verification_errors and flow through the existing sbom.signature_verification warning. Focused library and release-policy tests cover the error and exact operator-facing result, and the generated rule documentation was updated.

File: policy/lib/sbom/sbom.rego [hunk 17]
```diff
@@ -25,13 +34,20 @@ cyclonedx_sboms := [s | some s in _cyclonedx_sboms]
...
 
-_cyclonedx_sboms_from_input contains statement.predicate if {
-	some att in input.attestations
+_spdx_sboms := _spdx_sboms_without_image | _spdx_sboms_from_image
+_spdx_sboms_without_image := _spdx_sboms_from_attestations | _spdx_sboms_from_pipelinerun
+
+# Signed SBOM attestations carry the SBOM directly as the in-toto predicate.
+# Verification is shared by both format-specific rules below; predicateType
+# determines which result set receives each document.
+_cyclonedx_sboms_from_attestations contains statement.predicate if {
```
[the remark this answers is not in the document]
Comment (joejstuart) [reply policy/lib/sbom/sbom.rego:49]: No additional code change planned. Commit f2b08647 already documents the intentional public-contract change and migration in both design/sigstore-verification.md and example/data/rule_data.yml. Release/downstream announcement is coordination outside this policy change. A global missing-identity warning would be misleading for valid PipelineRun-only users.

File: example/data/rule_data.yml [hunk 18]
```diff
@@ -38,9 +38,11 @@ rule_data:
     rh-release:
       public_key: "k8s://openshift-pipelines/release-signing-key"
       ignore_rekor: true
-    # Add an "sbom" identity to gate SBOM discovery via OCI referrers and
-    # image-tag refs: SBOMs are only accepted when their signatures verify
-    # against it, and excluded (fail-closed) when it is absent. Left
+    # An "sbom" identity is required to expose attached SBOM attestations and
```
[the remark this answers is not in the document]
Comment (joejstuart) [reply example/data/rule_data.yml:41]: No change planned. The example comment already states that the identity is required, explains the prior behavior, and says absent or invalid verification fails closed. That is sufficient configuration guidance; the repository design document provides the deeper rationale, while an internal file-path pointer would not help consumers of packaged or copied rule data.

File: policy/lib/sbom/sbom.rego [hunk 19]
```diff
@@ -514,21 +548,39 @@ rule_data_errors contains error if {
...
+}
+
 # Collect SBOM signature verification errors for observability. Fail-closed
 # exclusion happens in the verified_* wrappers; this surfaces WHY an SBOM
 # was excluded, using the failures API to avoid duplicate builtin calls.
 signature_verification_errors contains error if {
+	_sbom_signing_identity_configured
+	verification := ec.sigstore.verify_attestation(input.image.ref, _sbom_signing_identity)
+	object.get(verification, "success", false) == false
+	some raw_error in verification.errors
```
[the remark this answers is not in the document]
Comment (joejstuart) [reply policy/lib/sbom/sbom.rego:570]: No change planned. The pinned ec.sigstore.verify_attestation builtin always returns an errors array on every success and failure path. Direct iteration is the documented pattern for its statically typed result. Defaulting a hypothetical missing field to [] would still emit no warning, so the proposed fixture would only model an invalid builtin response.

**Review by robnester-rh — APPROVED** [review]
LGTM

**Review by BohdanMar — APPROVED** [review]
Lgtm
